### PR TITLE
Add language subpaths to NextJs example

### PR DESCRIPTION
**Pull request by isaachinman**

### CODE REVIEW

**Comment by isaachinman** [comment]
@jamuhl Here's a possible solution to what we were discussing in #535.

This is an opinionated approach, but then again any approach would have to be. In general, here are the things we needed to do:

1. Allow NextJs to render identically for `/page1` and `/de/page1`
2. Allow i18n to understand (and set) the locale difference between `/page1` and `/de/page1` during SSR
3. Allow client to watch for all route changes, and make sure the incoming path contains the correct (if any) language subpath
4. Allow client to watch for all locale changes, and make sure the incoming path contains the correct (if any) language subpath

Generally speaking, the first two are server-oriented and the relevant changes are in `server.js`.

The second two are client-oriented and the relevant changes are in `pages/_app.js`, as this is the "entry point" for our client.

I've also added a config file, where you can define your default language, and all other languages.

The end result:

    /page1 (English)
    /de/page1 (German)
    /de (Redirects to /de/ to make parsing more dependable)

Changing the route via the NextJs router directly, or changing the locale via the i18n instance directly will both result in validation of the incoming route against current or incoming locale. It's important to note that I can just write a `<Link href="/page3">` or `Router.push("/page3")`, without having to worry about current language, and our middleware will set the correct localised path. Another nice added benefit is that locale changes are also persisted across browser history. If you change locales, you can hit the back button on your browser and go back to the previous locale.

Side note: I'm not really sure what is happening with `pages/page3.js`, but the h1 title breaks. I think this was something you already noticed previously.

Hopefully all of this makes sense. Let me know what you think. There are probably some improvements to be made here, but most likely I will be building a production application taking this approach shortly.

**Comment by jamuhl** [comment]
🙌 you get the "i18next hero of the month" badge - this is awesome

File: example/nextjs/server.js [hunk 1]
```diff
@@ -39,8 +61,34 @@ i18n
         // missing keys
         server.post('/locales/add/:lng/:ns', i18nextMiddleware.missingKeyHandler(i18n));
 
+        // Add language subpaths via i18n middleware
+        i18nextMiddleware.addRoute(
```
Comment (jamuhl) [inline example/nextjs/server.js]: is this needed? does not the * route below with the next.js handle work alone

Comment (isaachinman) [reply example/nextjs/server.js]: Yes, totally correct! Forgot about that block, I should have removed it.

**Comment by isaachinman** [comment]
Glad to help. Not sure if you think it's smart to merge this, because it does make the example a bit more complicated... If anyone has any questions about how it works they can contact me directly.

**Comment by isaachinman** [comment]
@jamuhl I've come across a strange SSR bug, where namespaces don't seem to load correctly, and I was wondering if you can help me debug it.

[I've created a branch on my fork here](https://github.com/isaachinman/react-i18next/tree/ssr-rendering-broken) that demonstrates the problem as simply as possible.

Commenting and uncommenting out [this line](https://github.com/isaachinman/react-i18next/blob/ssr-rendering-broken/example/nextjs/pages/index.js#L10) seems to break the loading of the `common` namespace for the entire app, despite `_app.js` having its own `NamespacesConsumer` which is supposed to be loading in the `common` namespace.

Moreover, even if I wrap `Navbar` in its own `withI18next(['common'])` HOC instead of taking `t` from props, our SSR is still broken.

Any idea what's going on here? My initial impression was that our app could contain as many `NamespacesConsumer` as we desired, and we could connect components arbitrarily to different namespaces via the `withI18Next` util from any point in the component tree.

I would appreciate any insight you can offer, I have a feeling my lack of understanding comes from lack of experience in the i18next ecosystem in general.

**Comment by isaachinman** [comment]
Tracked it down to [this function](https://github.com/isaachinman/react-i18next/blob/master/example/nextjs/i18n.js#L42).

Somehow, despite the React tree being rendered containing more than one `NamespacesConsumer`, calling for more than one namespace, the second argument of `i18n.getInitialProps` is a single namespace.

Any idea why? Is it possible this is a bug within `react-i18next`?

**Comment by jamuhl** [comment]
Will have to look deeper into this on Tuesday...won't have time tomorrow...my little girl has her birthday.

The problem - i guess is in the function you mention - the basic sample just uses an approach which expects the page level withI18n hoc to contain all namespaces.

There is a function you can pass down `reportNS` you might incorporate that into the withI18n hoc -> the functionality is like: https://github.com/i18next/react-i18next/issues/499

tests for it: https://github.com/i18next/react-i18next/blob/master/test/withNamespaces.spec.js#L72

**Comment by isaachinman** [comment]
Ah yeah, seems like a simple misunderstanding. I at first though `i18n.getInitialProps` was a built-in that tied into `NamespacesConsumer` via `react-i18next`.

So, the problem generally speaking is: we need a way to traverse a React component tree, and determine what locales are required to render that tree correctly.

The suggestion from @olivierpascal in #499 looks interesting, but the performance of `renderToString` is famously bad. Did @olivierpascal mean that `react-i18next` would do a `renderToString` _just_ to determine the namespaces? If `renderToString` is the only approach, it seems like that belongs in userland. No one should be calling `renderToString` twice per request.

Perhaps it is a strange idea, but what about using something like `react-test-renderer`, and then recursively checking the JSON structure for namespace props?

I imagine that this will be necessary pretty quickly, when using the approach in this PR/example to build out an application of decent size. The goal is to be able to declare namespace dependencies via a decorator on each individual component, and have reconciliation/SSR happen in the most efficient way possible.

**Comment by olivierpascal** [comment]
@isaachinman the code you saw in #499 is supposed to be a server-side code for SSR purpose. The `renderToString` is not specifically used to report namespaces. This is the `renderToString` you already use if you are doing SSR.

**Comment by isaachinman** [comment]
So, does the top-level `reportNS` give _all_ namespaces within a tree already?

**Comment by olivierpascal** [comment]
Yep

**Comment by isaachinman** [comment]
Oh, that's great news. Thanks, that should be all we need here.

**Comment by isaachinman** [comment]
@jamuhl Let me know if you had an idea about how to use `I18nextProvider` in our current NextJs situation, it's not clear to me.

**Comment by jamuhl** [comment]
The provider is not needed in first place -> the withI18n takes the role of the provider passing down the i18n instance: https://github.com/i18next/react-i18next/blob/master/example/nextjs/lib/withI18next.js#L12

**Comment by jamuhl** [comment]
@isaachinman did you had luck with using the reportNS function?

**Comment by isaachinman** [comment]
@jamuhl Unfortunately I haven't had a chance yet - very busy week. I think it's worthwhile to get it working for this example though, or lots of people will hit the same confusion. Will update here when I get a chance to look at it.

**Comment by Sergey2095** [comment]
The problem with this implementation is that the page reloads when it has to work on the client side.

**Comment by isaachinman** [comment]
@Sergey2095 That is correct, when switching from default to non-default locale there will be a document request. I believe this is due to the way NextJs intercepts Express requests.

I was aware of this whilst building out the functionality, but do not currently know a way around it. I'd be very appreciative of any advice anyone might have.

Although it is an annoyance, my general outlook is that locale switching is not some "high performance", frequent activity that should retain SPA strictly. In reality this PR is mostly about routing and SEO concerns.

Again, happy to accept any advice. It's most likely a trivial thing about how the `Router.events.on('routeChangeStart')` function block is working.

**Comment by isaachinman** [comment]
@jamuhl I have some time to look at this `reportNS` solution now. What did you have in mind? How are we going to compile the argument returned from `reportNS` across multiple HOCs and pass that into `getInitialProps`?

**Comment by jamuhl** [comment]
merged this and made it with a config setting to opt-out: https://github.com/i18next/react-i18next/blob/master/example/nextjs/config.js#L10

**Pull request merged**